File: pkg/dev_compiler/test/codegen/expect/syncstar_yield_test.js
dart_library.library('syncstar_yield_test', null, /* Imports */[
  "dart_runtime/dart",
  'dart/core',
  'expect'
], /* Lazy imports */[
], function(exports, dart, core, expect) {
  'use strict';
  let dartx = dart.dartx;
  function* foo1() {
    yield 1;
  }
  dart.fn(foo1, core.Iterable$(core.int), []);
  function* foo2(p) {
    let t = false;
    yield null;
    while (true) {
      a:
        for (let i = 0; dart.notNull(i) < dart.notNull(dart.as(p, core.num)); i = dart.notNull(i) + 1) {
          if (!dart.notNull(t)) {
            for (let j = 0; dart.notNull(j) < 3; j = dart.notNull(j) + 1) {
              yield -1;
              t = true;
              break a;
            }
          }
          yield i;
        }
    }
  }
  dart.fn(foo2, core.Iterable$(core.int), [dart.dynamic]);
  function* foo3(p) {
    let i = 0;
    i = dart.notNull(i) + 1;
    p = dart.notNull(p) + 1;
    yield dart.notNull(p) + dart.notNull(i);
  }
  dart.fn(foo3, core.Iterable$(core.int), [core.int]);
  function main() {
    expect.Expect.listEquals([1], foo1()[dartx.toList]());
    expect.Expect.listEquals([null, -1, 0, 1, 2, 3, 0, 1, 2, 3], foo2(4)[dartx.take](10)[dartx.toList]());
    let t = foo3(0);
    let it1 = t[dartx.iterator];
    let it2 = t[dartx.iterator];
    it1.moveNext();
    it2.moveNext();
    expect.Expect.equals(2, it1.current);
    expect.Expect.equals(2, it2.current);
    expect.Expect.isFalse(it1.moveNext());
    expect.Expect.isFalse(it1.moveNext());
    expect.Expect.isFalse(it2.moveNext());
    expect.Expect.isFalse(it2.moveNext());
  }
  dart.fn(main);
  // Exports:
  exports.foo1 = foo1;
  exports.foo2 = foo2;
  exports.foo3 = foo3;
  exports.main = main;
});

File: tests/language_2/syncstar_yield_test.dart
// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

// TODO(jmesserly): this is a copy of the language test of the same name,
// we can remove this copy when we're running against those tests.
import "package:expect/expect.dart";

Iterable<int> foo1() sync* {
  yield 1;
}

Iterable<int> foo2(p) sync* {
  bool t = false;
  yield null;
  while (true) {
    a:
    for (int i = 0; i < p; i++) {
      if (!t) {
        for (int j = 0; j < 3; j++) {
          yield -1;
          t = true;
          break a;
        }
      }
      yield i;
    }
  }
}

// p is copied to all Iterators from the Iterable returned by foo3.
// Also each iterator will have its own i.
Iterable<int> foo3(int p) sync* {
  int i = 0;
  i++;
  p++;
  yield p + i;
}

main() {
  Expect.listEquals([1], foo1().toList());
  Expect.listEquals(
      [null, -1, 0, 1, 2, 3, 0, 1, 2, 3], foo2(4).take(10).toList());
  Iterable t = foo3(0);
  Iterator it1 = t.iterator;
  Iterator it2 = t.iterator; //# copyParameters: ok
  it1.moveNext();
  it2.moveNext(); //# copyParameters: continued
  Expect.equals(2, it1.current);
  // TODO(sigurdm): Check up on the spec here.
  Expect.equals(2, it2.current); // //# copyParameters: continued
  Expect.isFalse(it1.moveNext());
  // Test that two `moveNext()` calls are fine.
  Expect.isFalse(it1.moveNext());
  Expect.isFalse(it2.moveNext()); //# copyParameters: continued
  Expect.isFalse(it2.moveNext()); //# copyParameters: continued
}
